src/DurableTask.AzureStorage/Partitioning/LeaseCollectionBalancer.cs

```
@@ -133,25 +133,44 @@ public async Task StopAsync()
                 return;
             }
 
-            if (this.takerTask != null)
+            // This is the graceful/cooperative shutdown task. If this Task completes,
+            // then the worker has completed all its pending work items and should willingly
+            // give away its ownership leases
+            Task shutdownTask = Task.Run(async () =>
             {
-                this.leaseTakerCancellationTokenSource.Cancel();
-                await this.takerTask;
-            }
+                if (this.takerTask != null)
+                {
+                    this.leaseTakerCancellationTokenSource.Cancel();
+                    await this.takerTask;
+                }
+
+                await this.ShutdownAsync();
+                this.shutdownComplete = true;
+
+                if (this.renewTask != null)
+                {
+                    this.leaseRenewerCancellationTokenSource.Cancel();
+                    await this.renewTask;
+                }
+
+                this.leaseTakerCancellationTokenSource = null;
+                this.leaseRenewerCancellationTokenSource = null;
+            });
 
```

I expect this approach will probably work in cases where lease rebalancing happens as part of a shutdown. I'm not 100% certain whether all the cases we've seen involve shutdowns, or whether some of them could involve lease rebalancing (without any shutdowns). Maybe this is sufficient for the cases we care about, but are there any other places in the code where we wait before releasing the ownership lease?

---

Unfortunately, I don't know if there are other such cases and, even if I assumed there were, I would not be sure how to handle them yet.

I do see there's a few places in the code where exceptions are ignored. I'm looking to add some logs there to try and give us more of a chance to tackle these issues in the future. Hopefully that will make it easier to reason about the cases you bring up.